test/external_tests.py

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python3
+
+# ------------------------------------------------------------------------------
+# This file is part of solidity.
+#
+# solidity is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# solidity is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with solidity.  If not, see <http://www.gnu.org/licenses/>
+#
+# (c) 2023 solidity contributors.
+# ------------------------------------------------------------------------------
+
+from argparse import ArgumentParser, Namespace
+import os
+from pathlib import Path
+import sys
+import subprocess
+
+EXTERNAL_TESTS_DIR = Path(__file__).parent / "externalTests"
+
+
+class ExternalTestNotFound(Exception):
+    pass
+
+
+def detect_external_tests() -> dict:
+    return {
+        file_path.stem: file_path
+        for file_path in Path(EXTERNAL_TESTS_DIR).iterdir()
+        if file_path.is_file() and file_path.suffix == ".sh"
+    }
+
+
+def display_available_external_tests(_):
+    print("Available external tests:")
+    print(*detect_external_tests().keys())
+
+
+def run_test_scripts(solc_binary_type: str, solc_binary_path: Path, tests: dict):
+    for test_name, test_script_path in tests.items():
+        print(f"Running {test_name} external test...")
+        subprocess.run(
+            [test_script_path, solc_binary_type, solc_binary_path],
+            check=True
+        )
+
+
+def run_external_tests(args: dict):
+    solc_binary_type = args["solc_binary_type"]
+    solc_binary_path = args["solc_binary_path"]
+
+    all_test_scripts = detect_external_tests()
+    if args["run_all"]:
+        run_test_scripts(solc_binary_type, solc_binary_path, all_test_scripts)
+    else:
+        selected_tests = args["selected_tests"]
+        if selected_tests:
+            unrecognized_tests = set(selected_tests) - set(all_test_scripts.keys())
+            if unrecognized_tests != set():
+                raise ExternalTestNotFound(
+                    f"External test(s) not found: {', '.join(unrecognized_tests)}"
+                )
+            run_test_scripts(
```

Comment:
Just noticed that without the return this will now always raise the `ExternalTestNotFound`. Will push the fix.

Reply:
See: https://github.com/ethereum/solidity/pull/14170